modules/openiddict/src/Volo.Abp.OpenIddict.AspNetCore/Volo/Abp/OpenIddict/Handlers/AttachAbpCustomChallengeErrors.cs

```diff
@@ -26,13 +24,9 @@ public ValueTask HandleAsync(OpenIddictServerEvents.ProcessChallengeContext cont
     {
         Check.NotNull(context, nameof(context));
 
-        var properties = context.Transaction.Properties[typeof(AuthenticationProperties).FullName!].As<AuthenticationProperties>();
```

Comment:
FYI, there's a built-in handler in both OpenIddict 3.x and 2.x that already does that for the custom parameters added to `AuthenticationProperties.Parameters` so this handler shouldn't be necessary (and the cool thing is that you can return strongly-typed JSON properties via `AuthenticationProperties.Parameters`) 😃

Reply:
hi @kevinchalet 

I didn't found such a handler in 3.x.

![image](https://user-images.githubusercontent.com/6908465/200114540-028dab4d-b4b9-44ca-b5c1-897aee246314.png)

Reply:
@maliming hey 😄 

`AuthenticationProperties` has two collections:
  - `AuthenticationProperties.Items` (which was called `AuthenticationProperties.Dictionary` in OWIN/Katana): it's an untyped collection that allows flowing specific properties between the user code and the handler. Some handlers persist them in things like cookies or tokens, some don't (in OpenIddict 4.x, we'll bring back the logic that persisted these properties in the tokens as in OpenIddict 1.x/2.x).

  - `AuthenticationProperties.Parameters`: it was introduced in ASP.NET Core 2.1 and is of type `IDictionary<string,object?>`. It's the one you need to use to return custom parameters via the built-in handler.

TL;DR, you'll basically want to use the newer constructor that takes 2 dictionaries:

```csharp
var properties = new AuthenticationProperties(
    properties: new Dictionary<string, string>
    {
        [OpenIddictServerAspNetCoreConstants.Properties.Error] = OpenIddictConstants.Errors.InvalidGrant,
        [OpenIddictServerAspNetCoreConstants.Properties.ErrorDescription] = nameof(SignInResult.RequiresTwoFactor)
    },
    parameters: new Dictionary<string, string>
    {
        ["userId"] = Guid.NewGuid().ToString("N"),
        ["twoFactorToken"] = Guid.NewGuid().ToString("N")
    });

return Forbid(properties, OpenIddictServerAspNetCoreDefaults.AuthenticationScheme);
```

Reply:
Thanks @kevinchalet 

I got that. 💯